File: algorithms/sorting/quick-sort/python/quick_sort.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""quick_sort.py

Quick sort is a popular sorting algorithm that follows the "divide and
conquer" strategy. It works by selecting a "pivot" element from the array
and partitioning the other elements into two sub-arrays according to whether
they are less than or greater than the pivot. The sub-arrays are then recursively
sorted.
"""

def main() -> None:
    """Driving code."""
    sample_list = [38, 27, 43, 3, 9, 82, 10]
    print("Original list:", sample_list)
    sorted_list = quick_sort(sample_list)
    print("Sorted list:", sorted_list)

def quick_sort(unsorted_list: list) -> list:
    """Sorts a list using the quick sort algorithm.

    Args:
        unsorted_list (list): The list to be sorted.

    Returns:
        list: The sorted list.
    """
    elements = len(unsorted_list)
    if elements < 2:
        return unsorted_list

    pivot = unsorted_list[0]
    left = [x for x in unsorted_list[1:] if x <= pivot]
    right = [x for x in unsorted_list[1:] if x > pivot]

    return quick_sort(left) + [pivot] + quick_sort(right)

if __name__ == "__main__":
    main()

File: algorithms/sorting/quick-sort/zig/quick_sort.zig
const quickSort = fn (unsortedList: []anytype) []anytype {
    const elements: usize = unsortedList.len;

    // Base case
    if (elements < 2) {
        return unsortedList;
    }

    var currentPosition: usize = 0; // Position of the partitioning element

    // Partitioning loop
    for (1 .. elements) |i| {
        if (unsortedList[i] <= unsortedList[0]) {
            currentPosition += 1;
            const temp = unsortedList[i];
            unsortedList[i] = unsortedList[currentPosition];
            unsortedList[currentPosition] = temp;
        }
    }

    const temp = unsortedList[0];
    unsortedList[0] = unsortedList[currentPosition];
    unsortedList[currentPosition] = temp; // Brings pivot to its appropriate position

    const left = quickSort(unsortedList[0..currentPosition]); // Sorts the elements to the left of pivot
    const right = quickSort(unsortedList[currentPosition+1..elements]); // Sorts the elements to the right of pivot

    return left ++ @[unsortedList[currentPosition]] ++ right; // Merging everything together
};
